app/src/main/java/pl/edu/pjwstk/s999844/shoppinglist/activities/MainActivity.kt

```diff
@@ -85,21 +86,44 @@ class MainActivity : AbstractShoppingActivity() {
 	@Suppress("UNUSED_PARAMETER")
 	fun onClickFloatingButton(view: View) = startActivity(Intent(baseContext, AddItemActivity::class.java))
 
-	private fun changeItemCallback(item: RequiredItem, change: Int) {
+	/**
+	 * Update item's amount.
+	 *
+	 * @param[item] what to change amount of
+	 * @param[change] how much it should change, `null` for deleting it
+	 */
+	private fun changeItemCallback(item: RequiredItem, change: Int?) {
 		val dbItem: RequiredItem = shoppingListDao.findById(item.id)
 			?: return
 
-		val newAmount: Int = dbItem.amount + change
-		if (newAmount <= 0) {
+		if (change == null) {
 			shoppingListDao.delete(dbItem)
 		} else {
-			dbItem.amount = newAmount
+			dbItem.amount = change.plus(dbItem.amount).coerceAtLeast(0)
 			shoppingListDao.update(dbItem)
 		}
 	}
 
 	private fun observeDatabaseChange(items: List<RequiredItem>) {
-		(binding.mainListRecyclerView.adapter as ShoppingListAdapter).setItems(items)
+		val zeroItemsComparator = Comparator<RequiredItem> { a, b ->
```

Comment:
We should probably only initialize the `zeroItemsComparator` in case the setting is active. Move it under the `if` below.

Reply:
moved it to an `if` which returns the chained `Comparator`s. the else branch feels a bit akward but that's just me.

Reply:
Instead of doing the following:
```kotlin
val orderComparator = CREATE_ORDER_COMPARATOR
val comparator = if (this.settings.orderZeroItemsLast) {
    val zeroComparator = CREATE_ZERO_COMPARATOR
    zeroComparator.then(orderComparator)
} else {
    orderComparator
}
val sorted = items.sortedWith(comparator)
```
we can do this:
```kotlin
var comparator = CREATE_ORDER_COMPARATOR
if (this.settings.orderZeroItemsLast) {
    comparator = CREATE_ZERO_COMPARATOR.then(comparator)
}
val sorted = items.sortedWith(comparator)
```

Reply:
Also we don't need the `this` there.

Reply:
ho, good idea, I'm much into functional so having mutable variables are really not into my realm. thanks for the hint.